SIP-31: Introduce `onClientRequest` entrypoint (#164)
MetaMask/SIPs: SIPS/sip-31.md

  enforces stricter request isolation at the API level.

By introducing `onClientRequest`, we ensure that Snaps can define handlers
exclusively accessible by the client, reducing potential attack vectors.

## Specification

This proposal introduces a new OPTIONAL handler function, `onClientRequest`,
which a Snap MAY implement.

### Language

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
"SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" written
in uppercase in this document are to be interpreted as described in [RFC
2119](https://www.ietf.org/rfc/rfc2119.txt).

### Snap Implementation

#### On Client Request

This specification introduces to the Snap Platform a dedicated handler that
Snaps MAY implement to process JSON-RPC requests originating exclusively from
the client.

Example:

```typescript
import { OnClientRequestHandler } from "@metamask/snaps-sdk";

export const onClientRequest: OnClientRequestHandler = async ({
  request,
}) => {
  const result = /* Handle `request` */;
  return result;
};
```

The type of the `onClientRequest` handler is:

```typescript
type OnClientRequestHandler = (
  args: OnClientRequestArguments,

> ) => Promise<OnClientRequestReturnValue>;

Reviewer: ```suggestion
) => Promise<OnClientRequestResponse>;
```

---
sip: 31
title: Client-only RPC Entrypoint
status: Draft
author: Daniel Rocha (@danroc)
created: 2025-03-19
---

## Abstract

This proposal introduces a new entrypoint, `onClientRequest`, allowing Snaps to

> expose a handler that can only be called by clients.

Reviewer 2: by _the client_. (singular)

While the request `origin` can currently be used to differentiate requests from
clients and dapps, this new entrypoint increases separation and minimizes
the risks associated with `origin` spoofing or request routing bugs.

## Motivation

Currently, Snaps rely on the `origin` field of incoming requests to

> differentiate whether a request originates from the client or an external dapp.

Reviewer 2: whether a request originates from the client, _another Snap,_ or an external dapp.

However, this approach has potential risks:

- **Origin Spoofing**: Bugs or vulnerabilities in request validation could

>   allow dapps to masquerade as clients.

Reviewer 2: as _the client_.

>   allow dapps to masquerade as the client.

Reviewer: I guess this somewhat depends on what kind of origin spoofing is possible. I would assume that the most likely case is not that a dapp can masquerade as the client, but instead as another dapp.

- **Unintended Exposure**: If a Snap processes requests without strict

>   validation, it may unintentionally expose to dapps methods that are meant

Reviewer 2: expose methods _to dapps or other Snaps_ that are meant only for _the client_.

  only for clients.

- **Cleaner Separation**: Having a dedicated entrypoint for client requests
  enforces stricter request isolation at the API level.

By introducing `onClientRequest`, we ensure that Snaps can define handlers

> exclusively accessible by clients, reducing attack vectors.

Reviewer 2: by _the client_, reducing _potential_ attack vectors.

## Specification

> This proposal introduces a new optional handler function, `onClientRequest`,

Reviewer 2: optional could be all-caps OPTIONAL

which a Snap MAY implement.

### Language

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
"SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" written
in uppercase in this document are to be interpreted as described in [RFC
2119](https://www.ietf.org/rfc/rfc2119.txt).

> ### `onClientRequest` Entrypoint

Author: Should we also define a new endowment?

Reviewer 2: We don't need it if this is restricted to pre-installed Snaps right?

Author: Yes, but should we limit this entrypoint to pre-installed Snaps?

Reviewer 2: Well I'm only speculating so maybe this is not relevant, but my thinking is: 

1. We only plan to implement `onClientRequest` in flows for pre-installed Snaps. We don't have any plans to do this with any 3rd party Snaps yet. 
2. The biggest risk, which may or may not be relevant, is if a 3rd party Snap spoofs itself as a pre-installed Snap (some 3rd party Snap pretends to be the ENS Resolver Snap, or MMI Snap, for example), and thus gains malicious access. 

So this is why I think we either need to introduce an endowment or make it only for pre-installed Snaps. I don't think we can skip both options. But would like someone else to validate or correct my thinking here...

Reviewer: Is there a use-case for `onClientRequest` beyond preinstalled Snaps? 

My assumption was that we wanted `onClientRequest` for non-standard JSON-RPC requests sent to Snaps in one way or another affiliated with MetaMask. Generally standardized JSON-RPC for everyone (third parties and preinstalled Snaps) should have be separate standardized handlers (e.g. `onProtocolRequest`).

Reviewer 2: OK so sounds like we are on board to make this only for pre-installed Snaps. Sounds good to me.

> ### `onClientRequest` Entrypoint

Reviewer: Can we use the same formatting for this section as https://metamask.github.io/SIPs/SIPS/sip-29 ?

Snaps MAY define an `onClientRequest` handler with the following signature:

```typescript

> type OnClientRequest = (request: JsonRpcRequest) => Promise<JsonRpcResponse>;

Reviewer: ```suggestion
type OnClientRequest = (request: JsonRpcRequest) => Promise<Json>;
```

```

#### Parameters

- `request: JsonRpcRequest` – A JSON-RPC request object.

#### Returns

> A `Promise<JsonRpcResponse>`, which resolves to a JSON-RPC response object.

Reviewer: ```suggestion
A `Promise<Json>`, which resolves to any valid JSON structure.
```

#### Behavior

> - Clients MUST ensure that they are the origin of a request before invoking

Reviewer 2: Can also make this singular. 

- The client MUST ensure that it is the origin of the request before invoking

> - The client MUST ensure that they are the origin of a request before invoking

Reviewer: Hmm, wondering if we wanna be more explicit about how to do this.

In practice we would probably check `origin === 'metamask'`, but that is of course an implementation detail.

Reviewer 2: Also realizing it should be "The client MUST ensure that **it is** the origin..."

Reviewer: ```suggestion
- The client MUST ensure that it is the origin of a request before invoking
```

Reviewer 2: Maybe be explicit by giving an example with the MetaMask client

  `onClientRequest`.

- Requests from other origins MUST be rejected.

- If a Snap does not implement `onClientRequest`, the client MUST throw an

>   exception for any request, regardless of the origin.

Reviewer 2: is this correct? any request? the Snap can still on `onRPCRequest` and other request interfaces right?

Author: This sentence isn't clear, I meant requests to the `onClientRequest` entrypoint. I just updated it.
